Make hardcoded parts of the optimizer sequence configurable (#13376)
ethereum/solidity: docs/internals/optimizer.rst

Selecting Optimizations
-----------------------

Detailed information regrading the optimization sequence as well a list of abbreviations is

> available in the :ref:`Yul optimizer docs <optimization-step-sequence>`.

Reviewer: This is the optimizer docs :) `yul.rst` documents the assembly (i.e. Yul):
```suggestion
available in the :ref:`assembly documentation <optimization-step-sequence>`.
```

Author: Not sure I get what you mean - was this wrong to begin with, because I've made no changes here? Also, this links to `yul.rst` `Optimization Step Sequence` anchor.

Reviewer: True. Looks like it was wrong before. Or maybe not exactly wrong - I see now that Yul docs just have a small section about the optimizer so it's technically correct - but a bit misleading since the main optimizer docs are actually here, not there.

Looking at it now, I think it's a bit weird to have that section with content that's not present in the main optimizer docs. I think we should move most of its content to the main optimizer docs and leave it just as a quick stub with a link.

.. code-block:: bash

    solc --optimize --ir-optimized --yul-optimizations 'dhfoD[xarrscLMcCTU]uljmul:fDnTOc'

The sequence inside ``[...]`` will be applied multiple times in a loop until the Yul code

> remains unchanged or until the maximum number of rounds (currently 12) has been reached.

Reviewer: I see that we have two sections about the sequence (here and in `yul.rst`). We should combine them into one and have one place link to the other.

Author: Describing everything in `optimizer.rst` and then linking from `yul.rst` would make the most sense to me in this case. Thoughts?

Reviewer: Seems like you did this the other way around :) But fine with me either way.

> The colon delimiter ``:`` is used to supply a custom clean up sequence in order to replace the

Reviewer: ```suggestion
The colon delimiter ``:`` is used to supply a custom clean-up sequence in order to replace the
```

Author: I'll refer to it as `cleanup`, as that's what it's been referred to throughout the doc; even the section describing it is called `Cleanup`. If you want me to change all instances to `clean-up`, please ping. Whatever we choose, I'd prefer to keep it consistent.

> The colon delimiter ``:`` is used to supply a custom clean up sequence in order to replace the

Reviewer: I think we should mention whether `:` is mandatory or not and whether omitting it is different than just leaving the part before/after it empty.

Also, I think we should mention that there are still some hard-coded steps before/after the sequences.

> default (``fDnTOc``) one, which is run after the stack compressor when using the legacy EVM

Reviewer: Let's not specify the default here. We won't remember to update it when it changes.
```suggestion
default one, which is run after the stack compressor when using the legacy EVM
```
Same in ``yul.rst``.

> code transform.

Reviewer: It sounds a bit as if the cleanup sequence was only used for the legacy code transform.

Author: True, I'll omit the mention of the stack compressor completely.
